src/connector/src/sink/encoder/mod.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::array::RowRef;
+use risingwave_common::catalog::Field;
+use risingwave_common::row::Row;
+
+use crate::sink::Result;
+
+mod json;
+
+pub use json::JsonEncoder;
+
+/// Encode a row of a relation into
+/// * an object in json
+/// * a message in protobuf
+/// * a record in avro
+pub trait RowEncoder {
```

Comment:
After rethinking our current sink implementation, I think instead of have a row level encoder, we can have a stream chunk encoder with the following definition
```
trait StreamChunkEncoder {
    fn encode(chunk: &StreamChunk) -> impl Iterator<(Vec<u8>, Vec<u8>)>;
}
```
which turns a stream chunk into an iterator that emits key-value pairs. Sinks that write data as key-value (e.g. kafka, kinesis, nats and redis) can reuse these encoders. In this way, even debezium can be an individual encoder that transform a stream chunk into a series of key-value pairs.

Reply:
1. Agree that `schema` and `col_indices` can be part of `encoder`
2. These will be handled by later part of the refactor, which is available for preview at #12125. That is, `Formatter` would look like an iterator (instead of current `Stream`), which is called repeatedly by `MessageSink` (kafka, kinesis, nats, redis).

Reply:
I have updated `encoder` to store `schema` and `col_indices`.